### PR TITLE
Expressions: Add ExprMacros.

### PR DESCRIPTION
**Pull request by gianm**
ExprMacros have the same syntax as functions but can convert themselves
to any kind of Expr at parse-time.

ExprMacroTable is an extension point for adding new ExprMacros. Anything
that might need to parse expressions needs an ExprMacroTable, which can
be injected through Guice.

The motivation here is two fold.

1. Provide a way to define functions where, for efficiency, some parts of the evaluation should be done ahead of time. The `like` function added here is an example: its pattern is compiled into a LikeMatcher object at expression-parse-time.
2. Provide a way for extensions to add new functions (they can do it through defining a macro).

### CODE REVIEW

**Comment by gianm** [comment]
Broken out from #4360.

**Comment by fjy** [comment]
👍

File: common/src/main/java/io/druid/math/expr/Expr.java [hunk 1]
```diff
@@ -347,7 +408,7 @@ protected final double evalDouble(double left, double right)
   @Override
   protected final long evalLong(long left, long right)
   {
-    return LongMath.pow(left, (int)right);
+    return LongMath.pow(left, (int) right);
```
Comment (leventov) [inline common/src/main/java/io/druid/math/expr/Expr.java]: Maybe `Ints.checkedCast()` for sanity

File: common/src/main/java/io/druid/math/expr/ExprListenerImpl.java [hunk 2]
```diff
@@ -285,15 +287,22 @@ public void exitPowOpExpr(ExprParser.PowOpExprContext ctx)
...
+                            ? (List<Expr>) nodes.get(ctx.getChild(2))
+                            : Collections.emptyList();
+
+    Expr expr = macroTable.get(fnName, args);
+
+    if (expr == null) {
+      // Built-in functions.
+      final Function function = Parser.getFunction(fnName);
+      if (function == null) {
+        throw new RuntimeException("function '" + fnName + "' is not defined.");
```
Comment (leventov) [inline common/src/main/java/io/druid/math/expr/ExprListenerImpl.java]: Could use Druid's `RE` with pattern.

File: common/src/main/java/io/druid/math/expr/ExprMacroTable.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+package io.druid.math.expr;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ExprMacroTable
+{
+  private static final ExprMacroTable NIL = new ExprMacroTable(Collections.emptyList());
```
Comment (leventov) [inline common/src/main/java/io/druid/math/expr/ExprMacroTable.java]: Why "nil"? Looks more like "empty"

File: common/src/main/java/io/druid/math/expr/ExprMacroTable.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  public static ExprMacroTable nil()
+  {
+    return NIL;
+  }
+
+  public Expr get(final String functionName, final List<Expr> args)
+  {
+    final ExprMacro exprMacro = macroMap.get(functionName.toLowerCase());
+    if (exprMacro == null) {
+      return null;
```
Comment (leventov) [inline common/src/main/java/io/druid/math/expr/ExprMacroTable.java]: Why silently consume the situation when the function is not found? Maybe throw exception?

File: processing/src/main/java/io/druid/query/expression/LikeExprMacro.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  public String name()
+  {
+    return "like";
+  }
+
+  @Override
+  public Expr apply(final List<Expr> args)
+  {
+    if (args.size() < 2 || args.size() > 3) {
+      throw new IAE("'%s' must have 2 or 3 arguments", name());
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/expression/LikeExprMacro.java]: `[%s]` is usually used

File: processing/src/main/java/io/druid/query/expression/LikeExprMacro.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      throw new IAE("pattern and escape must be literals");
+    }
+
+    final String escape = escapeExpr == null ? null : (String) escapeExpr.getLiteralValue();
+    final Character escapeChar;
+
+    if (escape != null && escape.length() != 1) {
+      throw new IllegalArgumentException("Escape must be null or a single character");
+    } else {
+      escapeChar = (escape == null || escape.isEmpty()) ? null : escape.charAt(0);
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/expression/LikeExprMacro.java]: `isEmpty()` check is redundant because length must be 1, it is checked above.

**Comment by leventov** [comment]
The transient integration test failure is: https://github.com/druid-io/druid/issues/4359#issuecomment-305863713

File: common/src/main/java/io/druid/math/expr/ExprMacroTable.java (hunk 3 above)
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprMacroTable.java]: I named it nil similar to how the empty list is named nil in Lisp and in Scala.

File: common/src/main/java/io/druid/math/expr/ExprMacroTable.java (hunk 4 above)
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprMacroTable.java]: In ExprListenerImpl, the macro table is checked first, then built-in functions. null return makes this pattern simpler; if this threw an exception we'd have to catch it and ignore it.

File: common/src/main/java/io/druid/math/expr/ExprListenerImpl.java (hunk 2 above)
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprListenerImpl.java]: Changed.

File: common/src/main/java/io/druid/math/expr/Expr.java (hunk 1 above)
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Expr.java]: Changed.

File: processing/src/main/java/io/druid/query/expression/LikeExprMacro.java (hunk 5 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/expression/LikeExprMacro.java]: Changed to `"Function[%s] must have 2 or 3 arguments"`

File: processing/src/main/java/io/druid/query/expression/LikeExprMacro.java (hunk 6 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/expression/LikeExprMacro.java]: Removed the redundant check.

File: common/src/main/java/io/druid/math/expr/ExprMacroTable.java (hunk 4 above)
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprMacroTable.java]: I annotated this with `@Nullable` and wrote a javadoc.

**Comment by gianm** [comment]
@leventov thanks for review; pushed changes.

File: common/src/main/java/io/druid/math/expr/Parser.java [hunk 7]
```diff
@@ -43,55 +42,44 @@
...
-    Supplier<Function> supplier = func.get(name.toLowerCase());
-    if (supplier == null) {
-      throw new IAE("Invalid function name '%s'", name);
-    }
-    return supplier.get();
+    FUNCTIONS = ImmutableMap.copyOf(functionMap);
   }
 
-  public static boolean hasFunction(String name)
+  public static Function getFunction(String name)
```
Comment (jihoonson) [inline common/src/main/java/io/druid/math/expr/Parser.java]: We need to check argument types maybe here for type safety in the future? I recognize nulls are currently returned for unsupported argument types, but if we're going to make Druid more SQL-compatible, we need to check types somewhere.

File: processing/src/main/java/io/druid/query/aggregation/DoubleMaxAggregatorFactory.java [hunk 8]
```diff
@@ -157,7 +161,9 @@ public String getName()
   @Override
   public List<String> requiredFields()
   {
-    return fieldName != null ? Collections.singletonList(fieldName) : Parser.findRequiredBindings(expression);
+    return fieldName != null
+           ? Collections.singletonList(fieldName)
+           : Parser.findRequiredBindings(Parser.parse(expression, macroTable));
```
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/aggregation/DoubleMaxAggregatorFactory.java]: It looks to be possible to keep parse result if ```fieldName``` is null in the constructor. Same for other ```AggregatorFactory```s.

**Comment by jihoonson** [comment]
@gianm thanks for breaking the patch into small ones. Looks good to me overall.

File: common/src/main/java/io/druid/math/expr/Expr.java [hunk 9]
```diff
@@ -20,14 +20,39 @@
...
+
+  /**
+   * Returns the value of expr if expr is a literal, or throws an exception otherwise.
+   *
+   * @return expr's literal value
+   *
+   * @throws IllegalStateException if expr is not a literal
+   */
+  @Nullable
+  default Object getLiteralValue()
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Expr.java]: I wanted to avoid having `instanceof` checks and casts all over the place.

File: common/src/main/java/io/druid/math/expr/Parser.java (hunk 7 above)
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Parser.java]: Yes, I think at some point, the expressions should be more aware of types. It will help with performance too. I think in the future, we want to be doing code generation and that will require advance knowledge of what type everything is going to be.

For now, they aren't type-aware.

File: processing/src/main/java/io/druid/query/aggregation/DoubleMaxAggregatorFactory.java (hunk 8 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/DoubleMaxAggregatorFactory.java]: It's true, but would require some refactoring of AggregatorUtil and its callers. I think it'd be better to do it in a separate patch.

File: common/src/main/java/io/druid/math/expr/Expr.java (hunk 9 above)
Comment (jihoonson) [inline common/src/main/java/io/druid/math/expr/Expr.java]: Ok.

File: common/src/main/java/io/druid/math/expr/Parser.java (hunk 7 above)
Comment (jihoonson) [inline common/src/main/java/io/druid/math/expr/Parser.java]: Thanks. It sounds good.

File: processing/src/main/java/io/druid/query/aggregation/DoubleMaxAggregatorFactory.java (hunk 8 above)
Comment (jihoonson) [inline processing/src/main/java/io/druid/query/aggregation/DoubleMaxAggregatorFactory.java]: Ok.

**Comment by gianm** [comment]
Thanks for the review @jihoonson.
